File: Python/ExampleTests/common/api_request_typed.py
from . import json_util as _json
from .api_request_string import ApiRequestString

class ApiRequestTyped:
    def __init__(self, method: str = '', page: str = '', parameter: str = '', body: str = ''):
        self.method = method
        self.page = page
        self.parameter = parameter
        self.body = body

    @classmethod
    def from_string_obj(cls, s: ApiRequestString) -> 'ApiRequestTyped':
        return cls(
            s.method,
            s.page,
            s.parameter,
            s.body
        )

    def to_string_obj(self) -> ApiRequestString:
        return ApiRequestString(
            str(self.method),
            str(self.page),
            str(self.parameter),
            str(self.body)
        )

    @staticmethod
    def to_string_list(items) -> list:
        return [t.to_string_obj() for t in items]

    @staticmethod
    def from_string_list(items) -> list:
        return [ApiRequestTyped.from_string_obj(s) for s in items]

    def to_json_value(self) -> dict:
        return {
            'Method': self.method,
            'Page': self.page,
            'Parameter': self.parameter,
            'Body': self.body,
        }

    def to_json(self) -> str:
        return _json.dumps(self.to_json_value())

    @classmethod
    def from_json_value(cls, m: dict) -> 'ApiRequestTyped':
        return cls(
            _json.as_str(_json.require(m, 'Method'), 'Method'),
            _json.as_str(_json.require(m, 'Page'), 'Page'),
            _json.as_str(_json.require(m, 'Parameter'), 'Parameter'),
            _json.as_str(_json.require(m, 'Body'), 'Body')
        )

    @classmethod
    def from_json(cls, text: str) -> 'ApiRequestTyped':
        return cls.from_json_value(_json.loads(text))

    @staticmethod
    def to_json_list(items) -> str:
        return _json.dumps([item.to_json_value() for item in items])

    @classmethod
    def from_json_list(cls, text: str) -> list:
        raw = _json.as_list(_json.loads(text), 'ApiRequestTyped')
        return [cls.from_json_value(e) for e in raw]

    def __str__(self):
        return (f'Method={self.method}' + ', ' +
                f'Page={self.page}' + ', ' +
                f'Parameter={self.parameter}' + ', ' +
                f'Body={self.body}')

    def _key(self):
        return (self.method, self.page, self.parameter, self.body)

    def __eq__(self, other):
        if not isinstance(other, ApiRequestTyped):
            return NotImplemented
        return self._key() == other._key()

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'ApiRequestTyped({self})'

File: TypeScript/ExampleTests/common/ApiRequestTyped.ts
import { ApiRequestString } from "./ApiRequestString.js";
import * as _json from "./json.js";

export class ApiRequestTyped {
  method: string;
  page: string;
  parameter: string;
  body: string;

  constructor(method: string, page: string, parameter: string, body: string) {
    this.method = method;
    this.page = page;
    this.parameter = parameter;
    this.body = body;
  }

  static fromStringObj(s: ApiRequestString): ApiRequestTyped {
    return new ApiRequestTyped(
      s.method,
      s.page,
      s.parameter,
      s.body
    );
  }

  toStringObj(): ApiRequestString {
    return new ApiRequestString(
      String(this.method),
      String(this.page),
      String(this.parameter),
      String(this.body)
    );
  }

  static toStringList(list: ApiRequestTyped[]): ApiRequestString[] {
    return list.map(t => t.toStringObj());
  }

  static fromStringList(list: ApiRequestString[]): ApiRequestTyped[] {
    return list.map(s => ApiRequestTyped.fromStringObj(s));
  }

  toJsonValue(): Record<string, unknown> {
    return {
      method: this.method,
      page: this.page,
      parameter: this.parameter,
      body: this.body,
    };
  }

  toJSON(): string { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m: unknown): ApiRequestTyped {
    return new ApiRequestTyped(
      _json.asString(_json.requireField(m, "method"), "method"),
      _json.asString(_json.requireField(m, "page"), "page"),
      _json.asString(_json.requireField(m, "parameter"), "parameter"),
      _json.asString(_json.requireField(m, "body"), "body")
    );
  }

  static fromJSON(text: string): ApiRequestTyped {
    return ApiRequestTyped.fromJsonValue(_json.parse(text));
  }

  static toJSONList(items: readonly ApiRequestTyped[]): string {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text: string): ApiRequestTyped[] {
    const raw = _json.asArray(_json.parse(text), "ApiRequestTyped") ?? [];
    return raw.map((e) => ApiRequestTyped.fromJsonValue(e));
  }

  toString(): string {
    return `Method=${this.method}, Page=${this.page}, Parameter=${this.parameter}, Body=${this.body}`;
  }

  equals(other: ApiRequestTyped): boolean {
    return this.method === other.method
      && this.page === other.page
      && this.parameter === other.parameter
      && this.body === other.body;
  }
}
